palettes.cpp

```diff
@@ -0,0 +1,186 @@
+#include <map>
+#include <algorithm>
+#include <vector>
+#include <random>
+
+#include "palettes.h"
+#include "common.h"
+#include "config.h"
+#include "configdialog.h"
+
+Palette::Palette(const std::array<Color, _PALETTE_SIZE> &colors) {
+	std::copy(colors.begin(), colors.end(), begin());
+}
+
+Palette::Palette(const std::initializer_list<Color> &i_list) {
+	std::copy(i_list.begin(), i_list.end(), begin());
+}
+
+const Palette *RandomPalettes::random(int rng_token) {
+	static std::map<int, Palette *> generated_palettes;
+
+	auto result = generated_palettes.find(rng_token);
+
+	if (result == generated_palettes.end()) {
+		generated_palettes[rng_token] = new_random_palette();
+	}
+
+	return generated_palettes.at(rng_token);
+}
+
+Palette *RandomPalettes::new_random_palette() {
+	static int bias_intensity = (int) (30.0f / ((cfg.at(MaxColors) + 14.0f) / config_ranges.at(MaxColors).second));
+
+	static float red_bias = rand() % bias_intensity * (rand() % 2 ? -1.0f : 1.0f);

```

Comment:
I'm aware that most of this math is really hard to read. Using the C++ standard library RNG may have helped the readability quite bit here but I HATE it, the interface and the performance are both absolute shit. So it's just gonna have to be read-only code for now.